ckanext/dalrrd_emc_dcpr/plugin.py

```diff
@@ -15,3 +21,83 @@ def update_config(self, config_):
 
     def get_commands(self):
         return [test_ckan_cmd]
+
+    # IAuthFunctions
+
+    def get_auth_functions(self) -> typing.Dict[str, typing.Callable]:
+        return {
+            "package_publish": authorize_package_publish,
+        }
+
+    def get_actions(self) -> typing.Dict[str, typing.Callable]:
+        return {
+            "package_create": package_create,
+            "package_update": package_update,
+            "package_patch": package_patch,
+        }
+
+
+def authorize_package_publish(
+    context: typing.Dict, data_dict: typing.Optional[typing.Dict] = None
+) -> typing.Dict[str, bool]:
+    # in here we can inspect the context and the data_dict in order to
+    # decide whether to authorize access or not
+    user_name = context.get("user")
+    owner_org = data_dict.get("owner_org") or data_dict.get("group_id")
```

Comment:
This seems ambiguous. Lets retrieve the organization where the package belongs. Groups should not be relevant to this IMO.

```suggestion
    owner_org = data_dict.get("owner_org")
```

Check the difference between organizations and groups here:

https://stackoverflow.com/a/21403607

Reply:
@ricardogsilva when creating dataset the `data_dict` doesnt contain `owner_org` property instead it has `group_id`, the ckan organization implementation is they are groups objects with a `is_organization` flag set to True, that why I used the `group_id` when the organization id is not available. CKAN also does in its codebase to counter the issue.

Reply:
Upon closer inspection of the ckan source code I'm going to ask that you provide a bit more context to the above statement of:

> when creating dataset the data_dict doesnt contain owner_org property instead it has group_id

looking at the implementation of [package_create](https://github.com/ckan/ckan/blob/ckan-2.9.4/ckan/logic/action/create.py#L50), I see that a package is created from the input dict. The resulting instance has an `owner_org` attribute, [which seems to be populated by direct retrieval](https://github.com/ckan/ckan/blob/ckan-2.9.4/ckan/model/domain_object.py#L129) of the `owner_org` key of the original data dict

I may be wrong, but it seems to me that, contrary to the above statement, when creating a dataset, the `data_dict` indeed has an `owner_org` key. The only case where it may not have it is when CKAN has been explicitly configured to allow creating packages outside of an organization. Either way, it doesn't seem like using the `group_id` is standard behavior.

Reply:
By creating the dataset I meant the whole process of adding a new dataset, which involve creating basic fields and adding the dataset resources and not a call to a single action in this case `package_create`. 

> looking at the implementation of [package_create](https://github.com/ckan/ckan/blob/ckan-2.9.4/ckan/logic/action/create.py#L50), I see that a package is created from the input dict. The resulting instance has an `owner_org` attribute, [which seems to be populated by direct retrieval](https://github.com/ckan/ckan/blob/ckan-2.9.4/ckan/model/domain_object.py#L129) of the `owner_org` key of the original data dict

In theory that is how it should work but it is not always the case.

> Either way, it doesn't seem like using the `group_id` is standard behavior.

Ok, so what should we do when the `data_dict` doesn't have any `owner_org` property?

Reply:
ooh @ricardogsilva I have come to understand what is happening! `data_dict` doesn't include the `owner_org` when called called from here https://github.com/kartoza/ckanext-dalrrd-emc-dcpr/blob/55ee21968e0eecfa6091bf867484d3d143a4dcbc/ckanext/dalrrd_emc_dcpr/templates/package/snippets/package_basic_fields.html#L70